[LTI Provider] Wiring up LTI views to the courseware template

mcgachey opened the pull request:
This change modifies the render_courseware view in the LTI Provider app to call the courseware.html template, using a set of options that suppress the edX chrome. For more details on the edX as an LTI provider feature, see the spec at https://docs.google.com/document/d/185hdPvIxcKtiDOLjb4sTGovA_WYXWz5Cd79gCzQwBms

This change needs the functionality from _get_module_by_usage_id in module_render.py. To avoid duplication, I've changed that method to be public. Is it still appropriate for it to be in the module_render.py file, or is there somewhere more generic where it would make sense to move it?

Tagging @ormsbee for feedback.

edx-webhook commented:
Thanks for the pull request, @mcgachey! I've created [OSPR-554](https://openedx.atlassian.net/browse/OSPR-554) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

ormsbee commented:
@cpennington: Review please?

singingwolfboy commented on lms/djangoapps/lti_provider/tests/test_views.py:
```diff
@@ -23,6 +23,38 @@
+}
+
+
+ALL_PARAMS = dict(list(LTI_DEFAULT_PARAMS.items()) + list(COURSE_PARAMS.items()))
```
Nitpick: this works, but it's a little difficult to follow the logic, since there are so many transformations. It would be clearer to do:
```python
ALL_PARAMS = {}
ALL_PARAMS.update(LTI_DEFAULT_PARAMS)
ALL_PARAMS.update(COURSE_PARAMS)
```

ormsbee replied:
As long as we're in nitpick territory, just dropping the list() works too ;-) :

```python
ALL_PARAMS = dict(LTI_DEFAULT_PARAMS.items() + COURSE_PARAMS.items())
```

singingwolfboy commented on lms/djangoapps/lti_provider/tests/test_views.py:
```diff
@@ -170,11 +181,91 @@ def test_forbidden_if_session_incomplete(self):
+
+
+MODULE_INSTANCE = MagicMock()
+MODULE_INSTANCE.render = MagicMock(return_value='Fragment')
```
Nitpick: you could also do
```python
MODULE_INSTANCE.render.return_value = "Fragment"
```

singingwolfboy commented on lms/djangoapps/lti_provider/tests/test_views.py:
```diff
@@ -170,11 +181,91 @@ def test_forbidden_if_session_incomplete(self):
+    Tests for the render_courseware method
+    """
+
+    def test_valid_launch(self, _key_mock, _course_mock, _access_mock, _module_mock, _render_mock):
```
To avoid having all of these unused arguments passed to each test function, it might be better to define all these patches in a setUp method:
```python
def setUp(self):
    super(RenderCoursewareTest, self).setUp()
    p1 = patch(patch('lti_provider.views.render_to_response', return_value='Rendered page')
    p1.start()
    self.addCleanup(p1.stop)
    p2 = patch('lti_provider.views.get_module_by_usage_id', return_value=(MODULE_INSTANCE, None))
    p2.start()
    self.addCleanup(p2.stop)
    # etc...
```
If you want to be fancy, you could even define some kind of mapping of dotpath to return value and iterate over that, to avoid duplicated code in this setUp method. But that might make it more difficult to follow, so that's a judgement call. You should do whatever you think is clearer.

singingwolfboy commented on lms/djangoapps/lti_provider/views.py:
```diff
@@ -149,4 +154,26 @@ def render_courseware():
+        'disable_footer': True,
+        'disable_tabs': True,
+        'staff_access': staff,
+        'xqa_server': settings.FEATURES.get('USE_XQA_SERVER', 'http://xqa:server@content-qa.mitx.mit.edu/xqa'),
```
I'm a little wary about having `mit.edu` hardcoded as a default, even if it's intended to be overridden. People in the Open edX community might enable this feature accidentally without configuring the URL, and there's a risk that it will just work -- which sounds like a good thing, until you realize that it could potentially put a lot of extra load on MIT's service, and MIT may not be prepared to support backwards-compatibility for this service forever. It would be better to have this raise an exception if this feature is turned on without a valid URL for this setting, with a clear, friendly exception message explaining what needs to be done to turn this service on. That's the route that we took for the [Harvard's `annotation_storage_url` setting](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/videoannotation_module.py#L51-L56).

mcgachey replied:
You're right - I wasn't terribly happy about it either. I was following the pattern from the courseware views file, but I was hoping that somebody had a better idea. I'll switch it to an example.com URL, and track down where it's used in order to make the error meaningful.

sarina replied:
I think the XQA server is not really much used anymore, either -- it was an experimental feature added in mid-2012 and I don't think maintained since then. I would much prefer to add some sort of setting to determine this url.

@carsongee could I get your input here -- what's the story behind the xqa server? Do you guys actively use/maintain this server? If it's still in use, would it be possible to have someone factor out all references to the server instead of having the url hardcoded in our codebase?

mcgachey commented:
Thanks for the feedback - I've pushed an updated patch. I changed the XQA server URL to example.com, and added text to the error message when the connection fails. It'd be good to get rid of the requirement to pass a string there (or at least make it conditional on the feature setting), but I'm hesitant to make any unnecessary changes to the templates before Andrew takes a pass over them.

sarina commented:
I think the larger more global change to how XQA stuff is defined should probably be separate from your work; it's just something you've uncovered. Plus I'd like to understand from @carsongee  how MIT is using the XQA server and who would be responsible for refactoring and testing that.

carsongee commented:
I don't think we are using it to my knowledge.  A DNS query chain leads down a rabbit hole to an undefined record:
```shell
#[10:46:41] cgee@atomos:~ $
dig content-qa.mitx.mit.edu +short
ichuang.mitx.mit.edu.
ichuang.m.i4x.org.
#[10:46:45] cgee@atomos:~ $
dig ichuang.mitx.mit.edu +short
ichuang.m.i4x.org.
#[10:46:49] cgee@atomos:~ $
dig ichuang.m.i4x.org +short
```
so the DNS that the default is set to doesn't exist, and a look at our settings doesn't show us overriding it anywhere.  Maybe @ichuang could shed some light?

sarina commented:
@carsongee @ichuang it sounds like perhaps the XQA infrastructure could perhaps be removed? I don't see the value in continuing to add it to new features if it's not being used. How can we capture that work as a to-do?

sarina commented:
@singingwolfboy were you thinking something like this for supplying a default url for xqa? https://github.com/edx/edx-platform/pull/7836

ichuang commented:
The XQA server is currently offline, but there is a plan to bring this
service back into use.

On Tue, Apr 28, 2015 at 1:51 PM, Sarina Canelake <notifications@github.com>
wrote:

> @singingwolfboy <https://github.com/singingwolfboy> were you thinking
> something like this for supplying a default url for xqa? #7836
> <https://github.com/edx/edx-platform/pull/7836>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/7818#issuecomment-97152194>.
>

singingwolfboy commented:
@sarina: yes, that seems reasonable to me

cpennington commented on lms/djangoapps/lti_provider/tests/test_views.py:
```diff
@@ -23,6 +23,38 @@
+    """
+    request = RequestFactory().post('/')
+    request.user = UserFactory.create()
+    request.user.is_authenticated = MagicMock(return_value=authenticated)
```
Nit: `request.user.is_authenticated.return_value = authenticated`

cpennington replied:
Oh, no, I take it back. I forgot the entire user wasn't being mocked. Carry on.

sarina commented:
@singingwolfboy @ormsbee you are still reviewing this? Sorry I derailed the conversation. Looks like Phil has responded to the original feedback.

singingwolfboy commented:
This looks good to me. :+1:

cpennington commented:
:+1:

sarina commented:
I'll merge this but will ask DB or Cale to verify on stage next week.
